File: models/health_bar_cnn.py
import torch.nn as nn
import torch

# CNN Model for Health Bar Detection
class HealthBarCNN(nn.Module):
    def __init__(self):
        super(HealthBarCNN, self).__init__()
        self.conv1 = nn.Conv2d(3, 32, kernel_size=3, stride=1, padding=1)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1)
        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
        self.fc1 = nn.Linear(64 * 16 * 16, 128)  # Adjust dimensions based on input image size
        self.fc2 = nn.Linear(128, 1)  # Output is the health percentage

    def forward(self, x):
        x = self.pool(torch.relu(self.conv1(x)))
        x = self.pool(torch.relu(self.conv2(x)))
        x = x.view(-1, 64 * 16 * 16)  # Flatten image tensor
        x = torch.relu(self.fc1(x))
        x = torch.sigmoid(self.fc2(x))  # Sigmoid to output health percentage (0-1)
        return x

File: models/health_bar_cnn.cpp
#include <torch/script.h>
#include <opencv2/opencv.hpp>
#include "health_bar_cnn.h"

// Inference function to predict health percentage
float predict_health(cv::Mat image) {
    torch::jit::script::Module module;
    try {
        // Load the CNN model (TorchScript model)
        module = torch::jit::load("models/saved/health_bar_cnn.pt");
    } catch (const c10::Error& e) {
        std::cerr << "Error loading the model\n";
        return -1;
    }

    // Convert OpenCV image to tensor
    torch::Tensor input_tensor = torch::from_blob(image.data, {1, image.rows, image.cols, 3}, torch::kByte);
    input_tensor = input_tensor.permute({0, 3, 1, 2});  // Convert to [N, C, H, W]
    
    // Forward pass
    torch::Tensor output = module.forward({input_tensor}).toTensor();
    
    return output.item<float>() * 100.0;  // Return health percentage
}

